src/validation.cpp

```diff
@@ -2813,33 +2848,62 @@ bool CChainState::InvalidateBlock(CValidationState& state, const CChainParams& c
         setDirtyBlockIndex.insert(invalid_walk_tip);
         setBlockIndexCandidates.erase(invalid_walk_tip);
         setBlockIndexCandidates.insert(invalid_walk_tip->pprev);
-        if (invalid_walk_tip->pprev == to_mark_failed && (to_mark_failed->nStatus & BLOCK_FAILED_VALID)) {
+        if (invalid_walk_tip == to_mark_failed->pprev && (to_mark_failed->nStatus & BLOCK_FAILED_VALID)) {
```

Comment:
I'm confused by this fix. Let's say we called `invalidateblock` on block 10. In that case `to_mark_failed` is block 10, `to_mark_failed->pprev` is block 9. Let's say we're almost done and we just disconnected the tip at height 11. Then `invalid_walk_tip` is block 11. So this comparison always returns `false`?

Reply:
If we're almost done, to_mark_failed got set to block 12 in the last iteration of the loop (yes, these variable names make no sense).

Reply:
Ah, it's actually `pindex` that stays at block 10 throughout this function. `to_mark_failed` is set to `invalid_walk_tip`. By the time we reach this line we chipped one block off of the tip.

Looking at the first iteration: let's say the original tip is block 20. `to_mark_failed` starts at block 10. We disconnect block 20 and mark it `BLOCK_FAILED_VALID`. The`if` statement here returns `false`, so block 20 incorrectly remains `BLOCK_FAILED_VALID`. The second iteration `to_mark_failed` points to block 20, so that when we set it to `BLOCK_FAILED_CHILD`.

Do I understand correctly that the reason for this "we fix it the next iteration" is because we might get shut down? So we need to wait with `BLOCK_FAILED_CHILD` until we're sure the parent is `BLOCK_FAILED_VALID `.

Reply:
Right, I mean its not *strictly* because we want to be safe in case of shutdown (though that's nice), and more that this hunk previously never did anything at all (though its intent was "descendants get marked FAILED_CHILD").